Sources/Basics/Archiver+Libzip.swift

```diff
@@ -0,0 +1,469 @@
+                fileSystem: self.fileSystem
+            )
+            guard let cancellationKey = self.cancellator.register(
+                name: "zipfile compressor for '\(directory)'",
+                handler: compressor
+            ) else {
+                throw StringError("cancellation")
+            }
+
+            DispatchQueue.sharedConcurrent.async {
+                defer { self.cancellator.deregister(cancellationKey) }
+                completion(.init(catching: {
+                    try compressor.compress()
+                }))
+            }
+        } catch {
+            return completion(.failure(error))
+        }
+    }
+
+    public func validate(path: AbsolutePath, completion: @escaping (Result<Bool, Error>) -> Void) {
+        do {
+            guard self.fileSystem.exists(path) else {
+                throw FileSystemError(.noEntry, path)
+            }
+
+            let validator = Libzip.Validator(
+                archivePath: path,
+                fileSystem: self.fileSystem
+            )
+
+            return try completion(.success(validator.validate()))
+        } catch {
+            return completion(.failure(error))
+        }
+    }
+
+    public func cancel(deadline: DispatchTime) throws {
+        try self.cancellator.cancel(deadline: deadline)
+    }
+}
+
+enum Libzip {
+    struct Extractor: Cancellable {
+        let archivePath: AbsolutePath
+        let destinationPath: AbsolutePath
+        let fileSystem: FileSystem
+        let cancelled = ThreadSafeBox(false)
+
+        init(
+            archivePath: AbsolutePath,
+            destinationPath: AbsolutePath,
+            fileSystem: FileSystem
+        ) {
+            self.archivePath = archivePath
+            self.destinationPath = destinationPath
+            self.fileSystem = fileSystem
+        }
+
+        func extract() throws {
+            let archive = try ReadonlyZipArchive(path: self.archivePath)
+            for entry in try archive.entries() {
+                guard !self.cancelled.get(default: false) else {
+                    continue
+                }
+                guard let entryName = entry.name else {
+                    print("** skipping entry \(entry.index), unknown name")
+                    continue
+                }
+                print("** extracting \(entryName)")
+                let entryDestinationPath = try destinationPath.appending(RelativePath(validating: entryName))
+                // FIXME: windows?
+                if entryName.suffix(1) == "/" {
+                    try fileSystem.createDirectory(entryDestinationPath)
+                } else {
+                    let content = try entry.readAll()
+                    try fileSystem.writeFileContents(entryDestinationPath, data: content)
+                }
+            }
+        }
+
+        func cancel(deadline: DispatchTime) throws {
+            self.cancelled.put(true)
+        }
+    }
+
+    struct Compressor: Cancellable {
```

Comment:
As it stores references to `FileSystem` and `ThreadSafeBox`, this probably has reference semantics and should be a `final class`.